### PR TITLE
feat: schema metadata API support

### PR DESCRIPTION
**Pull request by domodwyer**
This PR adds client support for the IOx [Schema API][api].

The new client method operates in terms of tables:

```go
func (c *Client) GetSchema(ctx context.Context, namespace string, table string) (map[string]ColumnType, error)
```

I made this choice because I believe this is how @wolffcm intends to use it - unfortunately he is on holiday and I'm going away tomorrow, so I took a bit of a guess! If we decide to expose this as a per-namespace API call instead, it's a simple code change :+1:

Currently the IOx schema API operates in terms of namespaces (returning all schemas in the namespace) - I created https://github.com/influxdata/influxdb_iox/issues/4997 to track adding a per-table API endpoint in IOx and perform client-side filtering to hide this implementation detail for now.

It's worth pointing out that this client is currently targeting different IOx services (ingester, router, querier) as if they were all reachable from a single gRPC connection / network address:

https://github.com/influxdata/influxdb-iox-client-go/blob/e67cd5cd332d07d149aea0fbcf0efceba9868d2f/iox_client.go#L41-L47

While this is true for the all-in-one running mode, in prod deployments the ingester / router / querier services will be running on distinct instances and not all APIs will be available on all nodes, including this schema API which is exposed only by the router.

[api]:https://github.com/influxdata/influxdb_iox/blob/40a852552045ff6929c1f31e7d2890a0b4e41f88/generated_types/protos/influxdata/iox/schema/v1/service.proto

---

* feat: support for schema API code generation (0a148d8)

      Modifies the gRPC / proto code generation helper script to generate both the
      existing ingester service code, and the schema service code.

* test: composable test client + write helpers (d383b87)

      Refactors the test helper code into two distinct responsibilities that can be
      composed together to build up more complex test cases:

      	* openNewDatabase: initialise an IOx client with a random database
      	* writeDataset: perform a write to a specified db/table

      Previously initialising a client also returned the write destination, which
      restricted tests to writing to a single (pre-determined) table / db.

* feat(metadata): schema API client (e67cd5c)

      Implements a GetSchema() metadata call that returns the set of columns and
      their data types for a specific table within a namespace.

      Redefines the column data types defined in the proto spec, decoupling callers
      from the generated wire format (internal) representation.

      Currently fetches all tables for a given namespace and filters the results for
      the desired table in the client - this should be replaced with a more
      efficient table-orientated schema lookup API once it is available in IOx:

      	https://github.com/influxdata/influxdb_iox/issues/4997

### CODE REVIEW

File: generate_influxdb_iox_protobuf.sh
Comment (domodwyer) [inline generate_influxdb_iox_protobuf.sh:55]: These `generate_service` calls will generate the proto glue for any service [listed here](https://github.com/influxdata/influxdb_iox/tree/main/generated_types/protos/influxdata/iox) :tada:

File: iox_client_meta_schema.go
Comment (domodwyer) [inline iox_client_meta_schema.go:30]: I chose to avoid using the proto-generated consts, to abstract the user of this client away from the wire-format / implementation details.

File: iox_client_meta_schema_test.go
Comment (domodwyer) [inline iox_client_meta_schema_test.go:40]: This shows an example of how this API is used by callers.

File: iox_client_test.go
Comment (domodwyer) [inline iox_client_test.go:45]: This change lets various helper funcs load the same values by sharing the same fallback value logic.

**Comment by domodwyer** [comment]
This API is currently sourcing the schemas from the catalog, so it will be strongly consistent and always return the current point-in-time snapshot of the table schema.

In the future, [we would like to use the router's schema cache][cache] which would be eventually consistent so I wanted to make sure this probable behavioural change is communicated. It will also have the advantage of being (much!) faster when the system is loaded.

[cache]:https://github.com/influxdata/influxdb_iox/issues/5007

File: iox_client_meta_schema.go
Comment (jacobmarble) [reply iox_client_meta_schema.go:30]: Great idea. We would also like to switch to FlightSQL at some point in the future, so abstractions now is 👍

File: iox_client_meta_schema_test.go
Comment (jacobmarble) [inline iox_client_meta_schema_test.go:24]: 👌

**Review by jacobmarble — APPROVED** [review]

**Comment by jacobmarble** [comment]
@wolffcm I'll wait for your review before merging.

File: iox_client_meta_schema.go
Comment (wolffcm) [inline iox_client_meta_schema.go:55]: Just to be clear, `namespace` here is the database name, which is `<org_id>_<bucket_id>`, right?

**Review by wolffcm — APPROVED** [review]
Looks fantastic. Really nice to come back from holiday and see this!

We have been getting the schema (specifically tag vs value vs timestamp columns) just via the `Schema` and `Metadata` methods on the Flight API. Should I expect that to stop working?

**Comment by alamb** [comment]
For the record, I think @domodwyer  is out for a while (weeks) so I am not sure how long it will take him to respond.

**Comment by domodwyer** [comment]
> We have been getting the schema (specifically tag vs value vs timestamp columns) just via the Schema and Metadata methods on the Flight API. Should I expect that to stop working?

I believe they will continue working (though I have a lot to catch up on after a long break), however I expect using the Schema API will provide lower latency responses. I am unsure of the consistency guarantees provided by the Flight API when operating in a distributed system (I shall investigate) but the router's Schema API should always provide the complete set of columns even in a distributed setup.

Currently each call to the router's Schema API will in turn cause the router to query the catalog (postgres), but in the future these schema queries will be executed against [in-memory state only][cache] which should result in orders of magnitude faster responses than the current flight APIs.

[cache]: https://github.com/influxdata/influxdb_iox/issues/5007

File: iox_client_meta_schema.go
Comment (domodwyer) [reply iox_client_meta_schema.go:55]: Yes sorry - an extra mental model conversion when switching between IOx/everything else - namespace is indeed a `org_bucket`, though it is not constrained to that format.

**Pull request merged**
